src/ir/selector_access_path_test.cc

```diff
@@ -0,0 +1,140 @@
+#include "selector_access_path.h"
+#include "field_selector.h"
+#include "selector.h"
+#include "src/common/testing/gtest.h"
+
+#include "absl/hash/hash_testing.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_split.h"
+
+#include <string>
+#include <vector>
+
+namespace raksha::ir {
+
+// Selector Access paths are easier to read and write in their string form, so
+// this is very useful for creating readable test inputs. This could possibly
+// be useful as a method in SelectorAccessPath itself, but, it's a bit
+// simplified right now, as it only splits on '.', the field separator
+// character. Also, I haven't found a use for it in non-test code yet.
+static absl::StatusOr<SelectorAccessPath> MakeSelectorAccessPathFromString(
+    const std::string str) {
+  const std::vector<std::string> selector_strs =
+      absl::StrSplit(str, '.', absl::SkipEmpty());
+
+  if (selector_strs.empty()) {
+    return absl::InvalidArgumentError(
+        "Expected a valid SelectorAccessPath string to have a"
+        "non-trivial leaf element.");
+  }
+
+  auto selector_rev_iter = selector_strs.rbegin();
+
+  // Create a leaf selector from the first selector string.
+  SelectorAccessPath selector_access_path =
+      SelectorAccessPath(Selector(FieldSelector(*selector_rev_iter)));
+  ++selector_rev_iter;
+
+  // Add all others as parents.
+  for (; selector_rev_iter < selector_strs.rend(); ++selector_rev_iter) {
```

Comment:
Is this kind of for loop better than the c++11 ones like `for(auto selector : selector_strs)` or is there a reason why this style of loop is needed? (Non-rhetorical though the C++11 style is more intuitive to me, where possible).

Reply:
There are two reasons why the regular range-based `for` syntax would not work here (or, at least, would not work without some additional mechanism which might look worse):

1. I'm treating the first element specially, using it to construct a leaf `SelectorAccessPath`. Of course, I could add an additional `bool first = true;` declaration outside the loop, use it to guard the stuff I'm doing before, and flip it to false after the first iteration.

2. I'm using the iterators returned by `rbegin()` and `rend()` to iterate through the vector in reverse order. The C++ range based for, of course, iterates over collections in standard order. You could imagine, of course, making a simple class that wraps a collection and has its `begin` and `end` be the inner collection's `rbegin` and `rend` respectively, but that might be more confusing than just using `rbegin` and `rend` directly.

The high level answer to your question: it's probably possible to use a range-based for in this example, but it would introduce its own less-standard readability challenges, so I'm not sure it's worth it.